File: mrbgems/mruby-binding/test/binding.rb
assert("Kernel.#binding") do
  assert_kind_of Binding, binding
end

assert("Binding#local_variables") do
  block = Proc.new do |a|
    b = 1
    binding
  end
  assert_equal [:a, :b, :block], block.call(0).local_variables.sort
end

assert("Binding#local_variable_set") do
  bind = binding
  1.times {
    assert_equal(9, bind.local_variable_set(:x, 9))
    assert_raise(NameError) { x }
    assert_equal([:bind, :x], bind.local_variables.sort)
  }
end

assert("Binding#local_variable_get") do
  bind = binding
  x = 1
  1.times {
    y = 2
    assert_equal(1, bind.local_variable_get(:x))
    x = 10
    assert_equal(10, bind.local_variable_get(:x))
    assert_raise(NameError) { bind.local_variable_get(:y) }
    assert_equal([:bind, :x], bind.local_variables.sort)
  }
end

assert("Binding#source_location") do
  skip unless -> {}.source_location

  bind, source_location = binding, [__FILE__, __LINE__]
  assert_equal source_location, bind.source_location
end

assert("Binding#dup") do
  x = 5
  bind1 = binding
  bind1.local_variable_set(:y, 10)
  bind2 = bind1.dup
  assert_equal 5, bind2.local_variable_get(:x)
  assert_equal 10, bind2.local_variable_get(:y)
  x = 50
  assert_equal 50, bind1.local_variable_get(:x)
  assert_equal 50, bind2.local_variable_get(:x)
  bind1.local_variable_set(:y, 20)
  assert_equal 20, bind1.local_variable_get(:y)
  assert_equal 20, bind2.local_variable_get(:y)
  bind1.local_variable_set(:z, 30)
  assert_raise(NameError) { bind2.local_variable_get(:z) }
  bind2.local_variable_set(:z, 40)
  assert_equal 30, bind1.local_variable_get(:z)
  assert_equal 40, bind2.local_variable_get(:z)
end

assert "Kernel#binding and .eval from C" do
  assert_raise(RuntimeError) { binding_in_c }
end

# The local-variable space a binding holds is a closed env carrying the
# special-variable slot past its locals (mruby/internal.h), and
# `local_variable_set` on a name the scope has not seen grows that stack
# through mrb_proc_merge_lvar(). The new locals take the ground the slot
# stood on, so the merge has to move it; an env sized without the slot,
# which out-of-tree code builds by hand, must not be read past its locals
# at all. The C helpers are in test/binding.c.
assert "Binding#local_variable_set moves the special-variable slot" do
  b = binding
  assert_true __binding_env_svar?(b)
  len = __binding_env_len(b)
  marker = "slot marker"
  assert_true __binding_env_slot_set(b, marker)

  b.local_variable_set(:merged_lvar, 42)
  assert_true __binding_env_svar?(b)
  assert_equal len + 1, __binding_env_len(b)
  assert_equal marker, __binding_env_slot_get(b)

  GC.start
  assert_equal 42, b.local_variable_get(:merged_lvar)
  assert_equal marker, __binding_env_slot_get(b)
end

assert "Binding#local_variable_set over an env without the slot" do
  b = binding
  len = __binding_env_len(b)
  assert_true __binding_env_drop_svar(b)
  assert_false __binding_env_svar?(b)

  b.local_variable_set(:merged_lvar, 42)
  assert_false __binding_env_svar?(b)
  assert_equal len + 1, __binding_env_len(b)

  GC.start
  assert_equal 42, b.local_variable_get(:merged_lvar)
end

assert "a binding taken in a `define_method` block sees the block's closure" do
  # `define_method` marks the block it installs a scope, the way `def` marks
  # a method body, but the block keeps the closure it was made with: the
  # locals a direct reference reaches from inside it are the binding's too.
  factory = Object.new
  def factory.build
    x = 10
    k = Class.new
    k.send(:define_method, :direct) { x }
    k.send(:define_method, :names) { binding.local_variables.sort }
    k.send(:define_method, :get) { binding.local_variable_get(:x) }
    k.send(:define_method, :set) { binding.local_variable_set(:x, 20) }
    k.send(:define_method, :own) { |a| b = 1; binding.local_variables.sort }
    k
  end
  k = factory.build
  o = k.new

  assert_equal [:k, :x], o.names
  assert_equal 10, o.get
  assert_equal [:a, :b, :k, :x], o.own(0)

  # The store reaches the captured local itself, so the direct reference and
  # the next instance both see it.
  assert_equal 20, o.set
  assert_equal 20, o.direct
  assert_equal 20, k.new.get

  # A `def` body carries no closure, and the locals around it are not its own.
  c = Class.new { def m; binding.local_variables; end }
  assert_equal [], c.new.m
end

File: mrbgems/mruby-binding/test/binding.c
#include <mruby.h>
#include <mruby/proc.h>
#include <mruby/internal.h>
#include <mruby/variable.h>

static mrb_value
binding_in_c(mrb_state *mrb, mrb_value self)
{
  return mrb_funcall_argv(mrb, mrb_obj_value(mrb->object_class), MRB_SYM(binding), 0, NULL);
}

/* The local-variable space a binding holds (binding_env_new_lvspace() in
   src/binding.c): a closed env born with the special-variable slot past
   its locals, and the env Binding#local_variable_set grows through
   mrb_proc_merge_lvar(). The four helpers below are what the shape of that
   stack can be asked about with; there is no Ruby-visible face for it. */
static struct REnv*
binding_env(mrb_state *mrb, mrb_value binding)
{
  mrb_value obj = mrb_iv_get(mrb, binding, MRB_SYM(env));

  mrb_check_type(mrb, obj, MRB_TT_ENV);
  return (struct REnv*)mrb_obj_ptr(obj);
}

static mrb_value
binding_env_svar_p(mrb_state *mrb, mrb_value self)
{
  return mrb_bool_value(MRB_ENV_SVAR_P(binding_env(mrb, mrb_get_arg1(mrb))));
}

static mrb_value
binding_env_len(mrb_state *mrb, mrb_value self)
{
  return mrb_int_value(mrb, MRB_ENV_LEN(binding_env(mrb, mrb_get_arg1(mrb))));
}

/* Rewrites the env into what out-of-tree code builds by hand: a closed env
   over a heap stack of exactly its locals, carrying no slot. The stack
   shrinks for real, so a later read one past the locals is an
   out-of-bounds access a sanitizer build reports. */
static mrb_value
binding_env_drop_svar(mrb_state *mrb, mrb_value self)
{
  struct REnv *e = binding_env(mrb, mrb_get_arg1(mrb));

  if (MRB_ENV_ONSTACK_P(e) || !MRB_ENV_SVAR_P(e)) return mrb_false_value();
  size_t len = (size_t)MRB_ENV_LEN(e);
  if (len == 0) return mrb_false_value();

  e->stack = (mrb_value*)mrb_realloc(mrb, e->stack, sizeof(mrb_value) * len);
  MRB_ENV_CLEAR_SVAR(e);
  return mrb_true_value();
}

/* Puts a marker in the slot and reads it back. What the core keeps there
   has no Ruby face of its own; any marked value stands in for it here,
   which is all the merge below moves. */
static mrb_value
binding_env_slot_set(mrb_state *mrb, mrb_value self)
{
  mrb_value binding, v;
  mrb_get_args(mrb, "oo", &binding, &v);

  struct REnv *e = binding_env(mrb, binding);
  if (!MRB_ENV_SVAR_P(e)) return mrb_false_value();
  MRB_ENV_SVAR_SLOT(e->stack, MRB_ENV_LEN(e)) = v;
  mrb_write_barrier(mrb, (struct RBasic*)e);
  return mrb_true_value();
}

static mrb_value
binding_env_slot_get(mrb_state *mrb, mrb_value self)
{
  struct REnv *e = binding_env(mrb, mrb_get_arg1(mrb));

  if (!MRB_ENV_SVAR_P(e)) return mrb_nil_value();
  return MRB_ENV_SVAR_SLOT(e->stack, MRB_ENV_LEN(e));
}

void
mrb_mruby_binding_gem_test(mrb_state *mrb)
{
  struct RClass *o = mrb->object_class;

  mrb_define_method(mrb, o, "binding_in_c", binding_in_c, MRB_ARGS_NONE());
  mrb_define_method(mrb, o, "__binding_env_svar?", binding_env_svar_p, MRB_ARGS_REQ(1));
  mrb_define_method(mrb, o, "__binding_env_len", binding_env_len, MRB_ARGS_REQ(1));
  mrb_define_method(mrb, o, "__binding_env_drop_svar", binding_env_drop_svar, MRB_ARGS_REQ(1));
  mrb_define_method(mrb, o, "__binding_env_slot_set", binding_env_slot_set, MRB_ARGS_REQ(2));
  mrb_define_method(mrb, o, "__binding_env_slot_get", binding_env_slot_get, MRB_ARGS_REQ(1));
}
